### PR TITLE
Issue #515 - Idioms.ConstructorInitializedMemberAssertion - fixing enum handling

### PR DESCRIPTION
**Pull request by Tungsten78**
The root cause is that CreateAnonymous can return the default enum value when setting up the specimen.  This leads to a false positive when asserting that the ctor value vs proptery/field are equal.

The approach is two fold:
* BuildSpecimenFromConstructor() skips the default value for enums
* In the event we're verifying an enum type with a single element associated to 0 (default value), ConstructorInitializedMemberException is always thrown with custom message due to our inability to discern whether initialization actually occurred

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,6 +5,7 @@ build/
 [Oo]bj/
 [Hh]elp/
 NuGetPackages/
+[pP]ackages
```
Comment (adamchester) [inline .gitignore]: Why is this added?

**Comment by adamchester** [comment]
Thank you for the investigation and PR on this issue! It looks like the build failed due to a problem out of our control with AppVeyor/Google. Are you able to force another build?

I will take a look in next couple of days.

**Comment by Tungsten78** [comment]
@adamchester, I'd be happy to force a build but unfortunately am not sure how to do this.

File: .gitignore (hunk 1 above)
Comment (Tungsten78) [inline .gitignore]: I meant to annotate this in my PR as an inquiry - sorry for the oversight.  The nuget packages were showing up as needing to be added following my first compilation of the fork. What is the accepted way to handle this?

Comment (Tungsten78) [inline .gitignore]: This was removed, the issue is that newer versions of NuGet are ignoring the repositoryPath redirect.  This happens to occur if using VS 2015 (Nuget 3.3) with restore packages on.  My apologies for not seeing the packages having been committed at the root level of the repo.

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs [hunk 2]
```diff
@@ -475,6 +474,84 @@ public void VerifyTypeWithPublicReadOnlyFieldsIncorrectlyInitialisedViaConstruct
             // Teardown
         }
 
+        [Fact]
+        public void VerifyWellBehavedEnumFieldDoesNotThrow()
+        {
+            // Fixture setup
+            var dummyComposer = new Fixture();
```
Comment (ploeh) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Why `dummyComposer`?

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs [hunk 3]
```diff
@@ -475,6 +474,84 @@ public void VerifyTypeWithPublicReadOnlyFieldsIncorrectlyInitialisedViaConstruct
...
+            // Exercise system and verify outcome
+            Assert.DoesNotThrow(() => sut.Verify(fieldInfo));
+            // Teardown
+        }
+
+        [Fact]
+        public void VerifyIllBehavedEnumFieldDoesThrow()
+        {
+            // Fixture setup
+            var dummyComposer = new Fixture();
```
Comment (ploeh) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Why `dummyComposer`?

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs [hunk 4]
```diff
@@ -475,6 +474,84 @@ public void VerifyTypeWithPublicReadOnlyFieldsIncorrectlyInitialisedViaConstruct
...
+            // Exercise system and verify outcome
+            Assert.Throws<ConstructorInitializedMemberException>(() => sut.Verify(fieldInfo));
+            // Teardown
+        }
+
+        [Fact]
+        public void VerifyWellBehavedEnumPropertyDoesThrow()
+        {
+            // Fixture setup
+            var dummyComposer = new Fixture();
```
Comment (ploeh) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Why `dummyComposer`?

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs [hunk 5]
```diff
@@ -475,6 +474,84 @@ public void VerifyTypeWithPublicReadOnlyFieldsIncorrectlyInitialisedViaConstruct
...
+            // Exercise system and verify outcome
+            Assert.DoesNotThrow(() => sut.Verify(propertyInfo));
+            // Teardown
+        }
+
+        [Fact]
+        public void VerifyIllBehavedEnumPropertyDoesThrow()
+        {
+            // Fixture setup
+            var dummyComposer = new Fixture();
```
Comment (ploeh) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Why `dummyComposer`?

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs [hunk 6]
```diff
@@ -475,6 +474,84 @@ public void VerifyTypeWithPublicReadOnlyFieldsIncorrectlyInitialisedViaConstruct
...
+            // Exercise system and verify outcome
+            Assert.Throws<ConstructorInitializedMemberException>(() => sut.Verify(propertyInfo));
+            // Teardown
+        }
+
+        [Fact]
+        public void VerifyDefaultOnlyEnumFieldDoesThrow()
+        {
+            // Fixture setup
+            var dummyComposer = new Fixture();
```
Comment (ploeh) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Why `dummyComposer`?

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs [hunk 7]
```diff
@@ -475,6 +474,84 @@ public void VerifyTypeWithPublicReadOnlyFieldsIncorrectlyInitialisedViaConstruct
...
+            // Exercise system and verify outcome
+            Assert.Throws<ConstructorInitializedMemberException>(() => sut.Verify(fieldInfo));
+            // Teardown
+        }
+
+        [Fact]
+        public void VerifyDefaultOnlyEnumPropertyDoesThrow()
+        {
+            // Fixture setup
+            var dummyComposer = new Fixture();
```
Comment (ploeh) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Why `dummyComposer`?

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs [hunk 8]
```diff
@@ -475,6 +474,84 @@ public void VerifyTypeWithPublicReadOnlyFieldsIncorrectlyInitialisedViaConstruct
...
+            var sut = new ConstructorInitializedMemberAssertion(dummyComposer);
+            var propertyInfo = typeof(ReadOnlyPropertyIncorrectlyInitializedViaConstructor<TestEnum>).GetProperty("Property");
+
+            // Exercise system and verify outcome
+            Assert.Throws<ConstructorInitializedMemberException>(() => sut.Verify(propertyInfo));
+            // Teardown
+        }
+
+        [Fact]
+        public void VerifyDefaultOnlyEnumFieldDoesThrow()
```
Comment (ploeh) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Could this test be refactored to a Parametrised Test and combined with `VerifyIllBehavedEnumFieldDoesThrow`?

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs [hunk 9]
```diff
@@ -475,6 +474,84 @@ public void VerifyTypeWithPublicReadOnlyFieldsIncorrectlyInitialisedViaConstruct
...
+            var sut = new ConstructorInitializedMemberAssertion(dummyComposer);
+            var fieldInfo = typeof(ReadOnlyFieldIncorrectlyInitializedViaConstructor<TestDefaultOnlyEnum>).GetField("Field");
+
+            // Exercise system and verify outcome
+            Assert.Throws<ConstructorInitializedMemberException>(() => sut.Verify(fieldInfo));
+            // Teardown
+        }
+
+        [Fact]
+        public void VerifyDefaultOnlyEnumPropertyDoesThrow()
```
Comment (ploeh) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Could this test be refactored to a Parametrised Test and combined with `VerifyIllBehavedEnumPropertyDoesThrow`?

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs [hunk 10]
```diff
@@ -705,5 +802,11 @@ int IEqualityComparer.GetHashCode(object obj)
                 throw new NotImplementedException();
             }
         }
+
+        enum TestEnum { none = 0, one, two, three };
+
+        enum TestDefaultOnlyEnum { none = 0 };
+
+        enum TestSingleNonDefaultEnum { none = 1 };
```
Comment (ploeh) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Where is this used?

**Comment by ploeh** [comment]
Thank you for your interest in contributing to AutoFixture. This pull request looks promising :+1: 

The few questions I had shouldn't be a major obstacle, I hope.

**Comment by adamchester** [comment]
It looks good.

I wonder have we thought if there are any other types that might have a similar issue? Types where we might not reliably be able to generate a non-default value? Enum is the only case I can think of.

Assuming the user actually has an enum with a single default value, now what can they do about it? I assume, given that we throw, they can no longer verify that class constructor automatically, and they have to manually test them? Would that manual test have to skip that parameter?

**Comment by Tungsten78** [comment]
Thanks both for the review.  I'll address each comment individually.

@adamchester - I too reflected on what other types might suffer this same fate and came up empty.  In regards to the single default value enum, I myself can't see the point of using an enum like this.  I felt the guard was important to prevent a false positive, but this seemed like a pretty rare scenario.  I agree with your solutions and would add another being possibly changing the value assignment if sticking to a single element was a requirement.

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 10 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: These are used as type arguments in the arrange phase of the newly added [unit tests](https://github.com/Tungsten78/AutoFixture/blob/905a836fe4248d40447e11b758ccfd01ae4f1ae6/Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs#L535).

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 8 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Great suggestion, I'll give it a try.

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 9 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Great suggestion, I'll give it a try.

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 7 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: To be honest, I was following the convention laid out by existing unit tests in this source file.  

Thinking about it, "dummy" might suggest a [dummy object](http://xunitpatterns.com/Dummy%20Object.html) which is not used by the test.  In this case, it is a critical stand-in for `ISpecimenBuilder`.

Any objections to `fixture` instead?   Shall I change all pre-existing instances as well? 
I assume we frown on inconsistency, but so to is mucking around too far outside the scope of a PR.

Comment (adamchester) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Looks like `dummyComposer` was my fault, and I think @ploeh was sick of my crappy PRs at the time and decided to allow it :)

Suggest we fix the names in a separate PR all together.

File: Src/Idioms/ConstructorInitializedMemberAssertion.cs [hunk 11]
```diff
@@ -250,11 +260,53 @@ public override void Verify(FieldInfo fieldInfo)
...
             ConstructorInfo ci, MemberInfo propertyOrField)
         {
             var parametersAndValues = ci.GetParameters()
-                .Select(pi => new {Parameter = pi, Value = this.builder.CreateAnonymous(pi)})
+                .Select(pi =>
+                {
+                    var value = this.builder.CreateAnonymous(pi);
+                    
+                    //ensure enum isn't getting the default value, otherwise we won't be able to determine whether initialization occurred
+                    if (pi.ParameterType.IsEnum && (int)value == default(int)) value = this.builder.CreateAnonymous(pi);
```
Comment (Tungsten78) [inline Src/Idioms/ConstructorInitializedMemberAssertion.cs]: I'd like to discuss this.

I was a bit uncomfortable doing this here thinking there might be a more declarative approach.  Perhapsa customization which overrides `EnumGenerator`'s behavior - say `EnumGeneratorNoDefault`.  Trouble is I wasn't comfortable moving this towards 'IFixture'.  Maybe there's a better way?

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 10 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Sorry, am not used to the comment vs line placement.
Thanks, `TestSingleNonDefaultEnum` is indeed missed in my tests.

**Comment by ploeh** [comment]
> I wonder have we thought if there are any other types that might have a similar issue?

Good question. `bool`?

I must admit that I can't even remember what the current algorithm is for `bool`, and I'm typing this on my way to another thing, so don't have time to look it up right now :flushed:

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 7 above)
Comment (ploeh) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: > Looks like `dummyComposer` was my fault, and I think @ploeh was sick of my crappy PRs at the time and decided to allow it :)

:laughing: 

> Suggest we fix the names in a separate PR all together.

:+1:

File: Src/Idioms/ConstructorInitializedMemberAssertion.cs (hunk 11 above)
Comment (ploeh) [inline Src/Idioms/ConstructorInitializedMemberAssertion.cs]: I looked at that line of code for quite a few minutes :smile: 

Then I thought: *oh, well, it's an implementation detail... If it's a problem, we can change it later.*

Not that I'm a proponent of not thinking things through, etc. but there's also the saying that *perfect is the enemy of good*. If you can think of a better way *right now*, then feel free to update the PR. If not, but you think of something next month, feel free to send another pull request :smile:

File: .gitignore (hunk 1 above)
Comment (Tungsten78) [inline .gitignore]: Pull Request #517 should help mitigate this in the future

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 2 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Addressed, now `fixture` in refactored tests.

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 3 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Addressed, now `fixture` in refactored tests.

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 4 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Addressed, now `fixture` in refactored tests.

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 5 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Addressed, now `fixture` in refactored tests.

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 6 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Addressed, now `fixture` in refactored tests.

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 7 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Addressed, now `fixture` in refactored tests.  

I will submit a PR that covers `dummyComposer` -> `fixture` in the context of this test file.

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 8 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Tests were refactored and are now relying on parameterised input.  The result is very nice.

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 9 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: Tests were refactored and are now relying on parameterised input. The result is very nice.

File: Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs (hunk 10 above)
Comment (Tungsten78) [inline Src/IdiomsUnitTest/ConstructorInitializedMemberAssertionTest.cs]: `TestSingleNonDefaultEnum` now incorporated into the parameterised tests

File: Src/Idioms/ConstructorInitializedMemberAssertion.cs (hunk 11 above)
Comment (Tungsten78) [inline Src/Idioms/ConstructorInitializedMemberAssertion.cs]: :thumbsup:

**Comment by Tungsten78** [comment]
I've been force pushing the PR branch to keep the commit clean.  I hope this is reasonable and not irritating.

Thanks for the feedback!

**Comment by adamchester** [comment]
:+1:

**Comment by ploeh** [comment]
Thank you for your contribution! It's now live as [AutoFixture 3.39.0](https://www.nuget.org/packages/AutoFixture/3.39.0).
